include/antworld/surface.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+// Standard C++ includes
+#include <vector>
+
+// OpenGL includes
+#include <GL/glew.h>
+
+// Libantworld includes
+#include "antworld/common.h"
+#include "antworld/opengl_type_traits.h"
+
+// Forward declarations
+namespace BoBRobotics
+{
+namespace AntWorld
+{
+class Texture;
+}
+}
+
+//------------------------------------------------------------------------
+// Surface
+//------------------------------------------------------------------------
+// Encapsulates a single 'surface' - geometry to be rendered in one draw call using one material
+namespace BoBRobotics
+{
+namespace AntWorld
+{
+class Surface
+{
+public:
+    Surface();
+    ~Surface();
+
+    //------------------------------------------------------------------------
+    // Public API
+    //------------------------------------------------------------------------
+    void bind(bool bindTexture = true) const;
+    void unbind(bool unbindTexture = true) const;
+    void unbindIndices() const;
+    void render(GLenum primitive = GL_TRIANGLES, GLenum indexType = GL_UNSIGNED_INT) const;
+
+    template<typename T>
+    void uploadPositions(const std::vector<T> &positions, GLint size = 3)
+    {
+        // Upload positions to buffer
+        uploadBuffer(positions, m_PositionVBO, GL_ARRAY_BUFFER, GL_STATIC_DRAW);
+
+        // Set vertex pointer and enable client state in VAO
+        glVertexPointer(size, OpenGLTypeTraits<T>::type, 0, BUFFER_OFFSET(0));
+        glEnableClientState(GL_VERTEX_ARRAY);
+
+        // Calculate number of vertices from positions
+        m_NumVertices = positions.size() / size;
+
+        // Unbind buffer
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+    }
+
+    template<typename T>
+    void uploadColours(const std::vector<T> &colours, GLint size = 3)
```

Comment:
If ``size`` can only be e.g. 1 and 3 we could enforce it with a ``BOB_ASSERT``

Reply:
OpenGL claims that it will produce a ``GL_INVALID_VALUE`` if these are invalid - our code is generally attaching an OpenGL error callback so I think this is probably ok

Reply:
Ok cool, as long as there's some error checking going on